jina/peapods/grpc/__init__.py

```diff
@@ -0,0 +1,151 @@
+        :param msg: the protobuf message to send
+        :param kwargs: Additional arguments.
+        """
+        routing_table = RoutingTable(msg.envelope.routing_table)
+
+        next_targets = routing_table.get_next_targets()
+
+        for target, _ in next_targets:
+            pod_address = target.active_target_pod.full_address
+
+            new_message = self._add_envelope(msg, target)
+
+            if pod_address not in self._stubs:
+                self._stubs[pod_address] = Grpclet._create_grpc_stub(pod_address)
+
+            try:
+                self.msg_sent += 1
+
+                # this wraps the awaitable object from grpc as a coroutine so it can be used as a task
+                # the grpc call function is not a coroutine but some _AioCall
+                async def task_wrapper(new_message, pod_address):
+                    await self._stubs[pod_address].Call(new_message)
+
+                asyncio.create_task(task_wrapper(new_message, pod_address))
+            except grpc.RpcError as ex:
+                self._logger.error('Sending data request via grpc failed', ex)
+                raise ex
+
+    @staticmethod
+    def send_ctrl_msg(pod_address: str, command: str):
+        """
+        Sends a control message via gRPC to pod_address
+        :param pod_address: the pod to send the command to
+        :param command: the command to send (TERMINATE/ACTIVATE/...)
+        :returns: Empty protobuf struct
+        """
+        stub = Grpclet._create_grpc_stub(pod_address, is_async=False)
+        response = stub.Call(ControlMessage(command))
+        return response
+
+    @staticmethod
+    def _create_grpc_stub(pod_address, is_async=True):
+        if is_async:
+            channel = grpc.aio.insecure_channel(
+                pod_address,
+                options=[
+                    ('grpc.max_send_message_length', -1),
+                    ('grpc.max_receive_message_length', -1),
+                ],
+            )
+        else:
+            channel = grpc.insecure_channel(
+                pod_address,
+                options=[
+                    ('grpc.max_send_message_length', -1),
+                    ('grpc.max_receive_message_length', -1),
+                ],
+            )
+
+        stub = jina_pb2_grpc.JinaDataRequestRPCStub(channel)
+
+        return stub
+
+    def _add_envelope(self, msg, routing_table):
+        new_envelope = jina_pb2.EnvelopeProto()
+        new_envelope.CopyFrom(msg.envelope)
+        new_envelope.routing_table.CopyFrom(routing_table.proto)
+        new_message = Message(request=msg.request, envelope=new_envelope)
+
+        return new_message
+
+    async def close(self, *args, **kwargs):
+        """Stop the Grpc server
+        :param args: Extra positional arguments
+        :param kwargs: Extra key-value arguments
+        """
+        self._logger.debug('Close grpc server')
+        await self._grpc_server.stop(grace=True)
```

Comment:
[`grace` needs to be an optional float value. ](https://grpc.github.io/grpc/python/grpc_asyncio.html#grpc.aio.Server.stop)